### DISCUSSION TITLE
Documents usage (Guide)

### DISCUSSION THREAD

**Discussion by c9482** [open-webui/open-webui#3432] [General]
Hello,

I am looking to start a discussion on how to use documents. Here's a starter question:

1. Is it more effective to use the model's Knowledge section to add all needed documents OR to refer to documents using # notation in the system prompt?

Many thanks in advance!

**Comment by derm1ch1** [comment]
Hi, I'm looking for more documentation on the use of documents as well. 
**To answer your question:** It is most effective to refer to a document with # since this gives the model a very specific idea where to look for information. I don't think this works in the system prompt though?
The models knowledge section is more convenient, and best if the user doesn't know which document contains the required information, the model will check a vector-db to find the right answer form all uploaded files (it's knowledge).


**To keep the discussion going**, I am wondering how the _"Workspace/ Documents"_ section and _"Admin Panel/ Settings/ Documents"_ are connected? 
I can't figure out if they are part of the same feature or seperate embedding-options?
The mentioned "Scan for documents from DOCS_DIR (/data/docs)" folder is completely empty for me, while I uploaded many documents to the "Documents" folder (which are available to the models). 

Does anyone have an answer or more information on this?
Many thanks in advance!

**Reply by c9482** [reply]
Actually, referring to a document with # notation in the system prompt works for me.

**Reply by Arijjebebli** [reply]
check /data/uploads

**Comment by ThibaudLopez** [comment]
> folder is completely empty for me, while I uploaded many documents to the "Documents" folder

The folder is at `/app/backend/data/docs/` .

---

To use RAG, the following steps worked for me (I have LLama3 + Open WebUI v0.3.5 Docker container):

1. I copied a `file.txt` from my computer to the Open WebUI container:
   `$ docker cp ~/Downloads/file.txt open-webui:/app/backend/data/docs/`

2. In the Open WebUI `Admin Panel` > `Settings` > `Documents` > `Scan for documents from DOCS_DIR (/data/docs)`, I clicked `Scan`. It confirmed with `Scan complete!`

3. I refreshed the `New Chat` page with `CTRL+R`.

4. In the prompt, I typed the hashtag `#` key, it listed the scanned files, I selected `file.txt`, I asked a question about it, for example _"what does the context contain?"_ , and it answered correctly about the contents of the file.

**Reply by mreitschuster** [reply]
why would the UI state

> Scan for documents from DOCS_DIR (/data/docs)

when the correct path is `/app/backend/data/docs/` ?

wouldnt a relative path make more sense (`./data/docs`), so the user has an indication he needs to consider the pwd?

Now that I found your explanation, it was easy to get it working. but i was so certain i had the right folder.

**Reply by Skorekaj** [reply]
I am following along your instructions , but i never see a SCAN button showing up , i have tried different browsers and tried to host the directories locally too but i never get the SCAN button when i docker cp myfile.pdf open-webui:/app/backend/data/docs

i'm user Ubuntu 22.04.5 LTS and Docker version 24.0.5

Any insights would be appreciated :)

**Comment by aman-knight** [comment]
how can i integrate any document parsing library in open web ui so i can parse my docs and feed to ollama models locally

**Comment by emaayan** [comment]
i'd like to join on this conversation, the whole process seems to be a little unclear, for example when i upload files like pdf and docx into the chat it completely ignores them  and i have to upload them into the document section, and click scan on each one,
then i have to open a new window . 

additionally if i remove them from the section, it doesnt' get removed from the mounted volume .

**Reply by Arijjebebli** [reply]
when i accessed their fast api there were endpoints for files and documents, u must delete it from there in order to have it remove it using their endpoint

**Comment by Vinay-052** [comment]
More information on embedding models usage, re-ranking models usage, defaulting them to GPU instead of CPU, support for adding \ using more embedding models is unclear, I see another discussion that using anything other than the default doesn't work and has been the same experience for me too. Even did a complete OS reinstall to check this and has been the same.

**Comment by tolumania** [comment]
Hii. I'm so grateful to have found this space and I'm really happy to just have access to help. So I pretty much just got introduced to open UI like 30 minutes ago. So I'm very clueless to a lot of things. I will definitely learn more as i keep going but i have an issue at hand now. I've not been able to get a successful response from open UI whenever i ask it something from a document i attach to it. It usually takes some time to answer even just normal questions from its memory which i also don't know what do concerning that btw. But it never responds when the document its supposed to give me the answer from is quoted directly. It just keeps loading and goes on forever. Sorry, I just want to know what can be done about that and also if there's anything i can add to the prompt to help. My prompt is empty and i don't know if that's responsible for some of this. I'll be very happy to get your input and answer any questions concerned with my models or settings and whatnot. Thank you so much

**Comment by tealcsolva** [comment]
I have just upgraded my stack and strugling to use RAG.

I created new 'knowledge' under Workspace and successfully uploaded 1 word doc. I can see the content through the UI and I can use # in prompt input as expected.
I create a new test prompt and mark the 'knowledge' by using #, but nothing happens. 
The UI shows the usual 'in progress' long lines and normally the output should come, but it just hangs.
I tried different settings, but no luck yet. I checked CPU usage on the backend and nothing happens after a sec, the CPU usage flat out to 0%.
The model gives output for non RAG prompts as expected.
I tested with a new 'knowledge' and uploaded one PDF and the problem occurs the same way.

open-webui-data/docs is empty
open-webui-data/uploads contains the docs I uploaded, starting with a hash+original name of file

Any hints why this is not working?

I read these resources about open web-ui:
https://docs.openwebui.com/tutorials/features/rag

Current stack:
ollama 0.3.12
model: llama3.1 8b
open webui: v0.3.32
- documents settings: Embedding model engine: default (SentenceTransformers)
- hybrid search: off
- Embedding model: sentence-transformers/all-MiniLM-L6-v2
- Content extraction engine: default
- RAG template: I've added some basic context inside the <context> tags. No special chars or anything.

**Comment by emaayan** [comment]
Dave  just posted a good vid exactly about this: https://www.youtube.com/watch?v=fFgyOucIFuk
